src/kernel/heap2.zig

```diff
@@ -0,0 +1,480 @@
+const std = @import("std");
+const Allocator = std.mem.Allocator;
+const testing = std.testing;
+const builtin = @import("builtin");
+const is_test = builtin.is_test;
+const build_options = @import("build_options");
+const mock_path = build_options.mock_path;
+const vmm = if (is_test) @import(mock_path ++ "vmm_mock.zig") else @import("vmm.zig");
+const log = @import("log.zig");
+const panic = @import("panic.zig").panic;
+
+const FreeListAllocator = struct {
+    const Error = error{TooSmall};
+    const Header = struct {
+        size: usize,
+        next_free: ?*Header,
+
+        const Self = @Self();
+
+        fn init(size: usize, next_free: ?*Header) Header {
+            return .{
+                .size = size,
+                .next_free = next_free,
+            };
+        }
+    };
+
+    first_free: ?*Header,
+    allocator: Allocator,
+
+    pub fn init(start: usize, size: usize) Error!FreeListAllocator {
+        if (size <= @sizeOf(Header)) return Error.TooSmall;
+        var first = @intToPtr(*Header, start);
+        first.* = Header.init(size - @sizeOf(Header), null);
+        return FreeListAllocator{
+            .first_free = first,
+            .allocator = .{
+                .allocFn = alloc,
+                .resizeFn = resize,
+            },
+        };
+    }
+
+    fn free(allocator: *Allocator, mem: []u8) void {
+        var self = @fieldParentPtr(FreeListAllocator, "allocator", allocator);
+        const addr = @ptrToInt(&mem[0]);
+        var header = @intToPtr(*Header, addr);
+        header.* = Header.init(mem.len - @sizeOf(Header), null);
+        if (self.first_free) |first| {
+            var prev: ?*Header = null;
+            // Find the previous free node
+            if (@ptrToInt(first) < addr) {
+                prev = first;
+                while (prev.?.next_free) |next| {
+                    if (@ptrToInt(next) > addr) break;
+                    prev = next;
+                }
+            }
+            // Make the freed header point to the next one, which is the one after the previous or the first if there was no previous
+            header.next_free = if (prev) |p| p.next_free else first;
+
+            // If there was a previous one, link it to the new one
+            if (prev) |p| {
+                p.next_free = header;
+            } else {
+                // Otherwise set the new one as the first one
+                self.first_free = header;
+            }
+
+            // Join with the next one until the next isn't a neighbour
+            while (header.next_free) |next| {
```

Comment:
Since I added joining to the left I can replace this while loop with an if statement, now that it is guaranteed all consecutive free nodes on the left will have been joined together.